File: Algorithms/SortingSeaching/BitonicSort.java

import java.util.*;

public class BitonicSort {

    private static void swap(int[] nums, int a, int b) {

        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;

    }

    private static void merge(int[] nums, int index, int length, boolean increasing) {

        if (length > 1) {

            int half = length / 2;

            for (int i = index; i < index + half; i++) {

                if ((nums[i] > nums[i + half] && increasing) ||
                    (nums[i] < nums[i + half] && !increasing)) {

                    swap(nums, i, i + half);

                }
            }

            merge(nums, index, half, increasing);
            merge(nums, index + half, half, increasing);

        }

    }

    public static void bitonicSort(int[] nums, int index, int length, boolean increasing) {

        if (length > 1) {

            int half = length / 2;

            bitonicSort(nums, index, half, true);
            bitonicSort(nums, index + half, half, false);

            merge(nums, index, length, increasing);

        }

    }

    public static void main(String[] args) {

        int[] nums = new int[]{1, 5, 77, 34, 2, 18, 19, 20, 21, 10, 43, 99, 101, 77, 75, 71}; // Must be pow of 2

        System.out.println(Arrays.toString(nums));

        bitonicSort(nums, 0, nums.length, true);

        System.out.println(Arrays.toString(nums));

    }

}

File: Algorithms/SortingSeaching/BitonicSort.py

def _merge(nums, index, length, increasing):

    if length > 1:

        half = length // 2

        for i in range(index, index + half):

            if ((nums[i] > nums[i + half] and increasing) or
                (nums[i] < nums[i + half] and not increasing)):

               nums[i], nums[i + half] = nums[i + half], nums[i]

        _merge(nums, index, half, increasing)
        _merge(nums, index + half, half, increasing)

def bitonic_sort(nums, index=None, length=None, increasing=None):

    if index == None or length == None or increasing == None:

        bitonic_sort(nums, 0, len(nums), True)

    else:

        if length > 1:

            half = length // 2

            bitonic_sort(nums, index, half, True)
            bitonic_sort(nums, index + half, half, False)

            _merge(nums, index, length, increasing)

def main():

    nums = [1, 5, 77, 34, 2, 18, 19, 20, 21, 10, 43, 99, 101, 77, 75, 71] # Must be pow of 2

    print(nums)
    bitonic_sort(nums)
    print(nums)

if __name__ == '__main__':
    main()
